### PR TITLE
perf(vault): virtualize host pickers for large inventories

### PR DESCRIPTION
**Pull request by binaricat**
## Summary

Virtualizes every remaining large host-picker surface so extreme inventories (thousands–tens of thousands of hosts) stay responsive. Select Host panel/dialog, Quick Switcher, SFTP host picker, Add to Workspace, and Create Workspace now render only the visible window instead of mapping every host into the DOM.

Also hardens virtual-list scroll clamping and keyboard cursors so empty search, filter-while-scrolled, and list shrink cannot blank the list or drive `selectedIndex` to `-1`.

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Refactor / code cleanup
- [ ] Documentation update
- [ ] Build / CI change
- [x] Other (please describe): Performance improvement

## Related Issue (optional)

N/A

## Changes Made

- Virtualize `SelectHostPanelContent` with flattened group/host rows + listbox keyboard model
- Virtualize `QuickSwitcher` categorized host/tab/shell/plugin rows
- Virtualize `SftpHostPicker` (local / connected / hosts)
- Virtualize `AddToWorkspaceDialog` and `CreateWorkspaceDialog` host lists
- Extract shared `virtualListMath` (scroll clamp, fixed window, keyboard index step)
- Apply render-path `effectiveScrollTop` clamp in Fixed/Variable virtual lists
- Add unit tests that drive shipped math (viewport slice, empty ArrowDown, visual index map)

## Screenshots / Demo

No visual redesign; same row chrome with virtualized scrolling under large host sets.

## Testing

- [ ] I have tested these changes locally (`npm run dev`)
- [x] Linting passes (`npm run lint`)
- [ ] Tests pass (`npm test`)
- [ ] Generated capability tool specs are updated when applicable (`npm run generate:capability-tools`)
- [x] No new console errors or warnings, if this affects app behavior

Focused tests (28/28 pass):

```
node --test --import tsx \
  components/ui/virtualListMath.test.ts \
  components/hostPickerVirtualization.test.ts \
  components/sftp/SftpHostPicker.test.ts \
  components/AddToWorkspaceDialog.keyboard.test.ts \
  components/QuickSwitcher.pluginContributions.test.tsx
```

ESLint on all changed TS/TSX files: clean (0 errors).

Two independent code reviews after fixes: CLEAN (no remaining blocking findings).

## Checklist

- [x] My code follows the existing project style
- [ ] I have added or updated relevant documentation
- [x] I have not introduced any breaking changes (or I have described them above)

### CODE REVIEW

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

File: components/QuickSwitcher.tsx [hunk 1]
```diff
@@ -294,70 +268,103 @@ const QuickSwitcherInner: React.FC<QuickSwitcherProps> = ({
...
+  }, [isOpen, itemIndexToVisualIndex, selectedIndex]);
+
+  const shellById = useMemo(
+    () => new Map(quickSwitcherShells.map((shell) => [shell.id, shell])),
+    [quickSwitcherShells],
+  );
+
+  const getRowHeight = useCallback(
+    (row: QuickSwitcherVisualRow) => (row.kind === "header" ? QS_HEADER_HEIGHT : QS_ROW_HEIGHT),
+    [],
```
[the remark this answers is not in the document]
Comment (binaricat) [reply components/QuickSwitcher.tsx:377]: Addressed in later commits: plugin command/view rows use `QS_PLUGIN_ROW_HEIGHT` (56px) via `getQuickSwitcherVisualRowHeight`, with unit coverage in `hostPickerVirtualization.test.ts`.

File: components/QuickSwitcher.tsx [hunk 2]
```diff
@@ -294,70 +270,106 @@ const QuickSwitcherInner: React.FC<QuickSwitcherProps> = ({
...
+    () => new Map(quickSwitcherShells.map((shell) => [shell.id, shell])),
+    [quickSwitcherShells],
+  );
+
+  const getRowHeight = useCallback((row: QuickSwitcherVisualRow) => {
+    if (row.kind === "header") return QS_HEADER_HEIGHT;
+    if (row.item.type === "plugin-command" || row.item.type === "plugin-view") {
+      return QS_PLUGIN_ROW_HEIGHT;
+    }
+    return QS_ROW_HEIGHT;
```
[the remark this answers is not in the document]
Comment (binaricat) [reply components/QuickSwitcher.tsx]: Addressed: all fixed-height Quick Switcher rows now use `h-full overflow-hidden` plus `truncate` / max-width constraints on host group paths, workspace titles, and session labels so long text cannot wrap into the next absolute virtual slot.

File: components/workspace/AddToWorkspaceDialog.tsx [hunk 3]
```diff
@@ -150,25 +197,92 @@ export const AddToWorkspaceDialog: React.FC<AddToWorkspaceDialogProps> = ({
...
+        className={`flex items-center justify-between px-4 py-2.5 cursor-pointer transition-colors ${getWorkspaceTargetRowStateClass(isCursor, isChecked)}`}
+        onClick={() => handleTargetClick(idx, host.id)}
+      >
+        <div className="flex items-center gap-3 min-w-0">
+          <DistroAvatar host={host} fallback={(host.label || host.hostname).slice(0, 2).toUpperCase()} size="sm" />
+          <span className="text-sm font-medium truncate">{host.label || host.hostname}</span>
+        </div>
+        <div className="flex items-center gap-2">
+          <div className="text-[11px] text-muted-foreground">
+            {host.group ? `Personal / ${host.group}` : 'Personal'}
```
[the remark this answers is not in the document]
Comment (binaricat) [reply components/workspace/AddToWorkspaceDialog.tsx]: Addressed: Add-to-Workspace host metadata uses `max-w-[12rem] truncate` (and the row is `overflow-hidden`) so long group paths stay inside the fixed 44px slot.

File: components/SelectHostPanelContent.tsx [hunk 4]
```diff
@@ -291,171 +544,61 @@ export const SelectHostPanelContent: React.FC<SelectHostPanelContentProps> = ({
...
+              <p>{t('selectHost.noHostsFound')}</p>
+            </div>
+          ) : (
+            <div
+              className="flex-1 min-h-0 px-3 py-2 outline-none"
+              data-host-picker-virtual="select-host"
+              role="listbox"
+              aria-label={t('selectHost.title')}
+              tabIndex={0}
+              onKeyDown={handleListKeyDown}
```
[the remark this answers is not in the document]
Comment (binaricat) [reply components/SelectHostPanelContent.tsx:619]: Addressed: the Select Host listbox now sets `aria-activedescendant` to a stable option id (`optionDomId(row.key)`), and each group/host option renders with a matching `id` so screen readers can announce the active virtualized row while focus stays on the listbox.

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review

**Comment by binaricat** [comment]
@codex review
